src/osd/scrubber/osd_scrub_sched.h

```diff
@@ -122,193 +122,483 @@ ScrubQueue interfaces (main functions):
- * scrub queue when all eligible PGs were already handled, or after a timeout
- * (or if their deadline has passed [[disabled at this time]]).
- */
-class ScrubQueue {
- public:
-  enum class must_scrub_t { not_mandatory, mandatory };
-
-  enum class qu_state_t {
-    not_registered,  // not a primary, thus not considered for scrubbing by this
-		     // OSD (also the temporary state when just created)
-    registered,	     // in either of the two queues ('to_scrub' or 'penalized')
-    unregistering    // in the process of being unregistered. Will be finalized
-		     // under lock
-  };
+  bool is_ripe(utime_t now_is) const
+  {
+    return urgency > urgency_t::off && !scrubbing && now_is >= not_before;
+  }
 
-  ScrubQueue(CephContext* cct, Scrub::ScrubSchedListener& osds);
-  virtual ~ScrubQueue() = default;
+  void update_ripe_for_sort(utime_t now_is)
+  {
+    eph_ripe_for_sort = is_ripe(now_is);
+  }
 
-  struct scrub_schedule_t {
-    utime_t scheduled_at{};
-    utime_t deadline{0, 0};
-  };
+  bool over_deadline(utime_t now_is) const
+  {
+    return urgency > urgency_t::off && now_is >= deadline;
+  }
 
-  struct sched_params_t {
-    utime_t proposed_time{};
-    double min_interval{0.0};
-    double max_interval{0.0};
-    must_scrub_t is_must{ScrubQueue::must_scrub_t::not_mandatory};
-  };
+  // status
+  void set_scrubbing()
+  {
+    scrubbing = true;
+    push_nb_out(5s);
+  }
+  void clear_scrubbing() { scrubbing = false; }
 
-  struct ScrubJob final : public RefCountedObject {
+  // failures
+  void push_nb_out(std::chrono::seconds delay);
+  void push_nb_out(std::chrono::seconds delay, delay_cause_t delay_cause);
+  void pg_state_failure();
+  void level_not_allowed();
+  void wrong_time();
+  void on_local_resources();
 
-    /**
-     *  a time scheduled for scrub, and a deadline: The scrub could be delayed
-     * if system load is too high (but not if after the deadline),or if trying
-     * to scrub out of scrub hours.
-     */
-    scrub_schedule_t schedule;
+  void dump(std::string_view sect_name, ceph::Formatter* f) const;
 
-    /// pg to be scrubbed
-    const spg_t pgid;
+  // consult the current value of the 'random upgrade" flag, and
+  // redraw the 'deep_or_upgraded' flag for the next run.
+  bool check_and_redraw_upgrade();
 
-    /// the OSD id (for the log)
-    const int whoami;
+  void set_oper_deep_target(scrub_type_t rpr);
+  void set_oper_shallow_target(scrub_type_t rpr);
 
-    ceph::atomic<qu_state_t> state{qu_state_t::not_registered};
+private:
 
-    /**
-     * the old 'is_registered'. Set whenever the job is registered with the OSD,
-     * i.e. is in either the 'to_scrub' or the 'penalized' vectors.
-     */
-    std::atomic_bool in_queues{false};
+  // updating periodic targets:
 
-    /// last scrub attempt failed to secure replica resources
-    bool resources_failure{false};
+  void update_as_shallow(
+      const pg_info_t& info,
+      const sched_conf_t& aconf,
+      utime_t now_is);
```

Comment:
I don't really understand why we are modifying ScrubTargets.  I'd expect to unschedule the existing one and then reschedule a new one.

Reply:
Why does it need access to both the pg_info_t and the sched_conf_t?  It seems to use stats_invalid and last_scrub_stamp.  Those seem like things that really should be in aconf?